fontforgeexe/charinfo.c

```diff
@@ -3909,28 +3984,81 @@ static void CIFillup(CharInfo *ci) {
     CI_DoHideUnusedPair(ci);
     CI_DoHideUnusedSingle(ci);
 
-    bits = SFGetAlternate(sc->parent,sc->unicodeenc,sc,true);
+	GGadget *cotf = GWidgetGetControl(ci->gw,CID_ComponentTextField);
+	GGadget *codcb = GWidgetGetControl(ci->gw,CID_ComponentDefaultCB);
+	GGadget *coim = GWidgetGetControl(ci->gw,CID_ComponentInterpMsg);
+	GGadget *cola = GWidgetGetControl(ci->gw,CID_ComponentChangeMsg);
+
+    bits = bits2 = SFGetAlternate(sc->parent,sc->unicodeenc,NULL,true);
     GGadgetSetTitle8(GWidgetGetControl(ci->gw,CID_ComponentMsg),
 	bits==NULL ? _("No components") :
 	hascomposing(sc->parent,sc->unicodeenc,sc) ? _("Accented glyph composed of:") :
 	    _("Glyph composed of:"));
     if ( bits==NULL ) {
 	ubuf[0] = '\0';
 	GGadgetSetTitle(GWidgetGetControl(ci->gw,CID_Components),ubuf);
+    GGadgetSetTitle(cotf,ubuf);
     } else {
-	unichar_t *temp = malloc(11*u_strlen(bits)*sizeof(unichar_t));
+	unichar_t *temp = malloc(18*u_strlen(bits)*sizeof(unichar_t));
 	unichar_t *upt=temp;
 	while ( *bits!='\0' ) {
-	    sprintf(buffer, "U+%04x ", *bits );
+	    sprintf(buffer, "U+%04x (", *bits );
 	    uc_strcpy(upt,buffer);
 	    upt += u_strlen(upt);
+	    if (iscombining(*bits)) {
+	        *upt = 0x25CC; // DOTTED CIRCLE “◌”
+	        upt += 1;
+	    }
+	    *upt = *bits;
+	    upt += 1;
+	    sprintf(buffer, ") ");
+	    uc_strcpy(upt,buffer);
+	    upt += u_strlen(upt);
+
 	    ++bits;
 	}
 	upt[-1] = '\0';
 	GGadgetSetTitle(GWidgetGetControl(ci->gw,CID_Components),temp);
 	free(temp);
+
     }
 
+    // i.e. five six hex digit long codepoints with spaces after them
+    char* codepoints_as_hex = malloc((6 * 5) + 5);
```

Comment:
Again, supposing it really maxes out, isn't this still missing space for the null terminator?